modules/carto/src/layers/cluster-utils.ts

```diff
@@ -0,0 +1,159 @@
+        out[parentId][name] = Math.max(prevValue, value);
+      } else if (aggregation === 'min') {
+        out[parentId][name] = Math.min(prevValue, value);
+      }
+    }
+  }
+
+  tile.userData[aggregationLevels] = Object.values(out);
+}
+
+export function extractAggregationProperties<FeaturePropertiesT extends {}>(
+  tile: Tile2DHeader<ParsedQuadbinTile<FeaturePropertiesT>>
+): AggregationProperties<FeaturePropertiesT> {
+  const properties: AggregationProperties<FeaturePropertiesT> = [];
+  const validAggregations: Aggregation[] = ['any', 'average', 'count', 'min', 'max', 'sum'];
+  for (const name of Object.keys(tile.content![0].properties)) {
+    let aggregation = name.split('_').pop()!.toLowerCase() as Aggregation;
+    if (!validAggregations.includes(aggregation)) {
+      aggregation = 'any';
+    }
+    properties.push({name: name as keyof FeaturePropertiesT, aggregation});
+  }
+
+  return properties;
+}
+
+export function computeAggregationStats<FeaturePropertiesT>(
+  data: ClusteredFeaturePropertiesT<FeaturePropertiesT>[],
+  properties: AggregationProperties<FeaturePropertiesT>
+) {
+  const stats = {} as Record<keyof FeaturePropertiesT, {min: number; max: number}>;
+  for (const {name, aggregation} of properties) {
+    stats[name] = {min: Infinity, max: -Infinity};
+    if (aggregation !== 'any') {
+      for (const d of data) {
+        stats[name].min = Math.min(stats[name].min, d[name] as number);
+        stats[name].max = Math.max(stats[name].max, d[name] as number);
+      }
+    }
+  }
+
+  return stats;
+}
+
+export function clustersToBinary<FeaturePropertiesT>(
+  data: ClusteredFeaturePropertiesT<FeaturePropertiesT>[]
+): BinaryFeatureCollection {
+  const positions = new Float32Array(data.length * 2);
+  const featureIds = new Uint16Array(data.length);
+  for (let i = 0; i < data.length; i++) {
+    positions.set(data[i].position, 2 * i);
+    featureIds[i] = i;
+  }
+
+  return {
+    shape: 'binary-feature-collection',
+    points: {
+      type: 'Point',
+      positions: {value: positions, size: 2},
+      properties: data,
+      numericProps: {},
+      featureIds: {value: featureIds, size: 1},
+      globalFeatureIds: {value: featureIds, size: 1}
+    },
+    lines: {
+      type: 'LineString',
+      positions: {value: new Float32Array(), size: 2},
+      pathIndices: {value: new Uint16Array(), size: 1},
+      properties: [],
+      numericProps: {},
+      featureIds: {value: new Uint16Array(), size: 1},
+      globalFeatureIds: {value: new Uint16Array(), size: 1}
+    },
+    polygons: {
+      type: 'Polygon',
+      positions: {value: new Float32Array(), size: 2},
+      polygonIndices: {value: new Uint16Array(), size: 1},
+      primitivePolygonIndices: {value: new Uint16Array(), size: 1},
+      properties: [],
+      numericProps: {},
+      featureIds: {value: new Uint16Array(), size: 1},
+      globalFeatureIds: {value: new Uint16Array(), size: 1}
+    } as any
```

Comment:
Are the empty line and polygon lists required?

Reply:
[Yes](https://github.com/visgl/deck.gl/blob/master/modules/layers/src/geojson-layer/geojson-layer.ts#L371-L372). The `GeoJsonLayer` crashes without them. I've at least made the code [more compact & reduced the number of needless allocations](https://github.com/visgl/deck.gl/pull/8957/commits/1501663d8f43b394d13d0bb4ecc83cb83317798e)